pkg/clusteragent/clusteragent.go

```diff
@@ -0,0 +1,127 @@
+package clusteragent
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+const (
+	defaultClusterAgentPort = "25000"
+	defaultTimeout          = 10 * time.Second
+)
+
+// Options should be used when initializing a new client.
+type Options struct {
+	// IgnoreNodeIPs is a set of ignored IPs that we don't want to pick for the cluster agent endpoint.
+	IgnoreNodeIPs sets.String
+	// Port overwrites the default cluster agent port to connect.
+	Port string
+	// InsecureSkipVerify skips the verification of the server's certificate chain and host name.
+	InsecureSkipVerify bool
+	// Timeout is the maximum amount of time a request is allowed to take.
+	Timeout time.Duration
+}
+
+type Client struct {
+	ip, port string
+	client   *http.Client
+}
+
+// NewClient picks an IP from one of the given machines and creates a new client for the cluster agent
+// with that IP.
+func NewClient(machines []clusterv1.Machine, opts Options) (*Client, error) {
+	var ip string
+	for _, m := range machines {
+		for _, addr := range m.Status.Addresses {
+			if !opts.IgnoreNodeIPs.Has(addr.Address) {
+				ip = addr.Address
+				break
+			}
+		}
+	}
+
+	if ip == "" {
+		return nil, errors.New("failed to find an IP for cluster agent")
+	}
+
+	port := defaultClusterAgentPort
+	if opts.Port != "" {
+		port = opts.Port
+	}
+
+	timeout := defaultTimeout
+	if opts.Timeout != 0 {
+		timeout = opts.Timeout
+	}
```

Comment:
Can we move these out of `NewClient` and instead make port and timeout necessary arguments? We can just pass the constant or custom value.